src/silly_sprites/global.hpp

```diff
@@ -0,0 +1,28 @@
+#include "event_system.hpp"
+
+namespace sly {
+    class Global final {
```

Comment:
I think the event system should be a member of `Application`. To provide access to other functions, we should define an interface to access various "global systems". `Application` should then implement that interface. I'd like to avoid having singletons if possible.

Reply:
is there an example for that concept already?
i dont quite get what you mean by that.

Reply:
The current implementation does not contain something like that (I think), but I wrote a little example here: https://godbolt.org/z/1T8cWEsrW
The benefits of such an architecture are:
* no global state
* no mutual dependencies between classes (`EventSystem` does not know about `Application`)
* only one common interface for all things that might be needed globally (`AppContext`)

But it also has drawbacks:
* a function expecting an `AppContext  const&` can access ***all*** of the provided "things", not only the needed ones (this could be mitigated by splitting the `AppContext` into a class hierarchy of its own, but I think that would be overkill)
* you always have to pass the context as a parameter when you need access to one the the "global" systems

Reply:
i removed the singleton but discovered a new problem.
when we want to have a RAII-Wrapper for the EventHandlerIds (as in the Test class in my `main.cpp`) we have no access to the application inside the dtor. every class/struct that neets to delete the ID in the dtor would the so store a ref to the application.
i recoment that every class should delete the ids inside the dtor so we have to invalid function pointer in the event system.

we could have one singleton that holds all global systems as the AppContext in TentakelsAttacking or do you have an other idea how to delete the ids inside the dtor?

Reply:
I think you could store a `AppContext*` inside the `EventHandlerId`s and then call `m_app_context->event_system()->remove_handler(m_value.value())` inside the destructor. For this to work, the `EventSystem` would also need an `AppContext*` as data member so that it can pass it to each newly created `EventHandlerId`.
Since the `Application` class implements the `AppContext` interface, it could just pass `this` to the constructor of `EventSystem` inside the constructor of `Application`. This is a bit dangerous, though: We have to ensure that `EventSystem` does not access any members of the `AppContext*` inside its constructor, because the `Application` object won't be fully created at that point in time.
I still think it's a better solution than creating a singleton or any other global state.

I think there are more ways to do this, but right now I think that the other alternatives that come to my mind would be worse than what I wrote above. But maybe you get another idea?